src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,289 @@
+  end,
+  mul_one := begin
+    rintros (i | i),
+    exact congr_arg a (add_zero i),
+    exact congr_arg xa (add_zero i),
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (i | i),
+    exact congr_arg a (neg_add_self i),
+    exact congr_arg a (sub_self (n + i)),
+  end }
+
+variable {n}
+
+@[simp] lemma a_mul_a (i j : zmod (2 * n)) : a i * a j = a (i + j) := rfl
+@[simp] lemma a_mul_xa (i j : zmod (2 * n)) : a i * xa j = xa (j - i) := rfl
+@[simp] lemma xa_mul_a (i j : zmod (2 * n)) : xa i * a j = xa (i + j) := rfl
+@[simp] lemma xa_mul_xa (i j : zmod (2 * n)) : xa i * xa j = a (n + j - i) := rfl
+
+lemma one_def : (1 : quaternion_group n) = a 0 := rfl
+
+private def fintype_helper : (zmod (2 * n) ⊕ zmod (2 * n)) ≃ quaternion_group n :=
+{ inv_fun := λ i, match i with
+                 | (a j) := sum.inl j
+                 | (xa j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := a j
+                 | (sum.inr j) := xa j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
+
+/-- The special case that more or less by definition `quaternion_group 0` is isomorphic to the
+infinite dihedral group. -/
+def quaternion_group_zero_equiv_dihedral_zero : quaternion_group 0 ≃* dihedral 0 :=
+{ to_fun :=  λ i, match i with
+                | (a j) := dihedral.r j
+                | (xa j) := dihedral.sr j
+                end,
+  inv_fun := λ i, match i with
+                | (dihedral.r j) := a j
+                | (dihedral.sr j) := xa j
+                end,
+  left_inv := by rintro (k | k); refl,
+  right_inv := by rintro (k | k); refl,
+  map_mul' := begin rintros (k | k) (l | l); try {refl},
+                    change quaternion_group_zero_equiv_dihedral_zero._match_1 (a (0 + l - k)) = _,
+                    rw zero_add,
+                    refl end }
+
+/-- Some of the lemmas on `zmod m` require that `m` is positive, as `m = 2 * n` is the case relevant
+in this file but we don't want to write `[fact (0 < 2 * n)]` we make this lemma a local instance. -/
+private lemma two_mul_pos_of_pos [fact (0 < n)] : fact (0 < 2 * n) :=
+begin
+rw ← zero_mul 0,
+          apply mul_lt_mul',
+            { exact zero_le 2 },
+            { assumption },
+            { exact zero_le 0 },
+            { exact nat.zero_lt_succ 1 }
+end
+
+local attribute [instance] two_mul_pos_of_pos
+
+/--
+If `0 < n`, then `quaternion_group n` is a finite group.
+-/
+instance [fact (0 < n)] : fintype (quaternion_group n) := fintype.of_equiv _ fintype_helper
+
+instance : nontrivial (quaternion_group n) := ⟨⟨a 0, xa 0, dec_trivial⟩⟩
+
+/--
+If `0 < n`, then `quaternion_group n` has `4n` elements.
+-/
+lemma card [fact (0 < n)] : fintype.card (quaternion_group n) = 4 * n :=
+begin
+  rw ← fintype.card_eq.mpr ⟨fintype_helper⟩,
+  change fintype.card (zmod (2 * n) ⊕ zmod (2 * n)) = 4 * n,
+  rw [fintype.card_sum, zmod.card, two_mul],
+  ring
```

Comment:
Does `abel` suffice?

(I feel like this proof should be easier, but I've lost my local oleans so can't investigate now.)

Reply:
After `abel` the goal is `4 •ℕ n = 4 * n` which I could finish with `rw [nsmul_eq_mul],
  norm_cast` which I am not sure is an improvement. I'll stick with ring for now, but feel free to suggest otherwise.